### PR TITLE
Tokenizing Input

### PR DESCRIPTION
**Pull request by seibelsabrina**
Changed the command line input processing to tokenize input before passing it to a command for execution.

PTAL

### CODE REVIEW

**Review by efuquen — COMMENTED** [review]
I think overall your style is fine, but the code is not working. I get the following error:

http://imgur.com/a/uNUHW

**Comment by efuquen** [comment]
Looks like none of the commands are recognized: http://i.imgur.com/F7oIEKM.png

Since `exit` isn't working I would take a look at what the `command` variable is being set to, seems like an issue with the `Tokenizer` if that is not getting set to the correct value:

https://github.com/lianasyrkett/CodeU-Summer-2017/pull/3/files#diff-085ee4c5309974341970845d11bcb286L58

File: src/codeu/chat/util/Tokenizer.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	private int at;
+
+	public Tokenizer(String source) {
+		token = new StringBuilder();
+		this.source = source;
+	}
+	public String next() throws IOException {
+		while (remaining() > 0 && Character.isWhitespace(peek())) {
+			read();
+		} if (remaining() <= 0) {
```
Comment (efuquen) [inline src/codeu/chat/util/Tokenizer.java]: Don't put the if statement right after the curly bracket of the `while` loop like this, it makes it confusing to read. The `if` statement should start on it's own line.

File: src/codeu/chat/util/Tokenizer.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+			return null;
+		} else if (peek() == '"') {
+			readWithQuotes();
+		} else {
+			readWithNoQuotes();
```
Comment (efuquen) [inline src/codeu/chat/util/Tokenizer.java]: You should be returning on each one of the branches of the `if` statement.

File: src/codeu/chat/util/Tokenizer.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		}
+		return "";
```
Comment (efuquen) [inline src/codeu/chat/util/Tokenizer.java]: If you return on each branch you shouldn't need this last return.

File: src/codeu/chat/util/Tokenizer.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	public Tokenizer(String source) {
+		token = new StringBuilder();
+		this.source = source;
+	}
+	public String next() throws IOException {
+		while (remaining() > 0 && Character.isWhitespace(peek())) {
+			read();
+		} if (remaining() <= 0) {
+			return null;
+		} else if (peek() == '"') {
```
Comment (efuquen) [inline src/codeu/chat/util/Tokenizer.java]: You should accept single quotes too, so add it to the conditional. Should be something like:

`} else if (peek() == '"' || peek() == '\'') {`

File: src/codeu/chat/util/Tokenizer.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		token.setLength(0);
+		while (remaining() > 0 && !Character.isWhitespace(peek())) {
+			token.append(read());
+		}
+		return token.toString();
+	}
+
+	private String readWithQuotes() throws IOException {
+		token.setLength(0);
+		if (read() != '"') {
```
Comment (efuquen) [inline src/codeu/chat/util/Tokenizer.java]: To accept single quotes do:
```
char openingQuote = read();
if (openingQuote != '"' && openingQuote != '\'') {
```

File: src/codeu/chat/util/Tokenizer.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+			throw new IOException("Strings must start with opening quote");
+		}
+		while (peek() != '"') {
```
Comment (efuquen) [inline src/codeu/chat/util/Tokenizer.java]: Similar to above, to accept single quotes do:
`while(peek() != '"' && peek() != '\'') {`

**Comment by seibelsabrina** [comment]
Did some small changes and the commands look like they are being recognized now. PTAL

**Comment by seibelsabrina** [comment]
Fixed Tokenizer.java as discussed in our meeting. Also resolved merge conflict. PTAL

File: src/codeu/chat/client/commandline/Chat.java [hunk 7]
```diff
@@ -308,7 +318,7 @@ private ConversationContext find(String title) {
     //
     panel.register("info", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
+      public void invoke(List<String> args) {
```
Comment (efuquen) [inline src/codeu/chat/client/commandline/Chat.java:321]: This is not compiling for me. Line 128 has an extra `panel.register("info", new Panel.Command() {` that is using the Scanner. This most likely came in when you merged in master to resolve the conflicts. I'm pasting the error below (:

```
src/codeu/chat/client/commandline/Chat.java:128: error: <anonymous codeu.chat.client.commandline.Chat$2> is not abstract and does not override abstract method invoke(List<String>) in Command
    panel.register("info", new Panel.Command() {
                                               ^
src/codeu/chat/client/commandline/Chat.java:129: error: method does not override or implement a method from a supertype
      @Override
      ^
```

**Review by efuquen — CHANGES_REQUESTED** [review]
When I removed the extra `info` command the code works like before but the quoting is not working properly. With single quotes it saves the whole user name with the quotes in it and double quotes cause the program to exit. output i get below:

```
>>> u-add efuquen
>>> u-list
USER efuquen (UUID:1.3793251597)
>>> u-add 'foobar'
>>> u-list
USER 'foobar' (UUID:1.2791704588)
USER efuquen (UUID:1.3793251597)
>>> u-add "blah"
Run PASSED
efuquen-macbookpro:CodeU-Summer-2017 efuquen$
```

**Pull request merged**